### ISSUE TITLE
Hyper MIO Branch Compatibility

### ISSUE THREAD

**Issue by bfrog** [iron/iron#446]
It would be great to see how, if at all, the mio branch of hyper will change things with Iron.

**Comment by bfrog** [comment]
One thing that I've found useful is the ability to have per listener thread state. In my case the Sender side of std::sync::mpsc::channel's, but I could see other per thread data being useful. So perhaps something akin to iron/persistent but with a way to set it up per worker thread. Maybe even simply allowing the user to create an iron server per thread as hyper does now would allow per thread state to be created for handlers.

**Comment by bfrog** [comment]
I spent some time investigating this and it really comes down to the Body readable at the moment is problematic. To avoid pushing all the async stuff into Iron the request body needs to either be entirely in memory or entirely on disk. As an alternative perhaps a continuation can be returned as a IronResult specifying it needs the Body to begin with. In all scenarios Body is no longer the same as it is in Iron 0.3

The in memory/on disk scenario I could see playing out basically where if the request body is extraordinarily large, its written to disk before being given. This is troublesome though because, what size is the maximum allowed? What's the cutoff for in memory or on disk?

The continuation method seems saner. Providing a closure as an IronResult, lets call it a BodyReader takes either the entirety or chunks of the body when needed...

So basically something like...

``` rust
fn some_handler(req: IronRequest) -> IronResult { 
     let request_handler = SomeHandler::new(req);
    IronResult::ReadBody(move |chunk| { request_handler.handle_chunk(chunk) }) 
}
```

Not quite as nice as the std::io::Read for sure, but without something akin to co-routines...

**Comment by reem** [comment]
So thinking about this more, the ideal situation is:

- we have an in-memory buffer for reading data from the socket
- we read into this buffer whenever a chunk is available until it is full
- user code can drain from this buffer asynchronously
- when the buffer is full we stop reading from the socket until the user code has drained some data, then we resume (the logic here is a bit tricky, but it should work)

This way we don't have to deal with an ever-expanding in-memory buffer or resorting to a disk buffer.

In terms of how we actually expose this to the user, something like a specialized `eventual` Stream would probably be the best way to go without going full coroutines, which could also work.

**Comment by reem** [comment]
In terms of how I'm thinking to land something like this: I'd like to do one more release of iron that will be the final release based off of sync hyper and that exposes a sync interface, then start another release for the asynchronous version. We can then maintain the synchronous version for some time before deprecating it.

**Comment by bfrog** [comment]
Seems like a sensible way of doing things to me

**Comment by yazaddaruvala** [comment]
@reem what are your thoughts on releasing a version of sync-iron where the backend is async-hyper?

Basically make it extremely easy for people to migrate over and test the new backend.

I guess the question is, do you think the cost of implementing a sync shim on top of hyper is low enough to warrant the added testing?

**Comment by Hoverbear** [comment]
With Hyper moving to `futures-rs` and `tokio` this issue is irrelevant. Hyper no longer has a `mio` branch, but it does have a `tokio` branch.

Closing this, if you're interested in improving the `async` situation please help out with Hyper's `tokio` branch. After it's ready we can get Iron working on it. :)

**Comment by untitaker** [comment]
We might as well rename this issue. :)

**Comment by KodrAus** [comment]
For discoverability I think it would be good to keep this issue around and open in some form until `iron` provides a `futures` based (or whatever it ends up being) implementation for asynchronous request handling.

**Comment by untitaker** [comment]
https://github.com/iron/iron/issues/501 here you go
